flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableDescriptor.java

```diff
@@ -0,0 +1,283 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Describes a {@link CatalogTable} representing a source or sink.
+ *
+ * <p>A {@link TableDescriptor} is a template for creating a {@link CatalogTable} instance. It
+ * closely resembles the "CREATE TABLE" SQL DDL statement, containing schema, connector options, and
+ * other characteristics. Since tables in Flink are typically backed by external systems, the
+ * descriptor describes how a connector (and possibly its format) are configured.
+ *
+ * <p>This can be used to register a table in the Table API, see (TODO FLINK-23065).
+ */
+@PublicEvolving
+public final class TableDescriptor {
+
+    private final Schema schema;
+    private final Map<String, String> options;
+    private final List<String> partitionKeys;
+    private final @Nullable String comment;
+
+    private TableDescriptor(
+            Schema schema,
+            Map<String, String> options,
+            List<String> partitionKeys,
+            @Nullable String comment) {
+        this.schema = Preconditions.checkNotNull(schema, "Table descriptors require a schema");
+        this.options = Collections.unmodifiableMap(options);
+        this.partitionKeys = Collections.unmodifiableList(partitionKeys);
+        this.comment = comment;
+    }
+
+    /**
+     * Creates a new {@link Builder} for a table using the given connector.
+     *
+     * @param connector The factory identifier for the connector.
+     */
+    public static Builder forConnector(String connector) {
+        final Builder descriptorBuilder = new Builder();
+        descriptorBuilder.option(FactoryUtil.CONNECTOR, connector);
+
+        return descriptorBuilder;
+    }
+
+    public Schema getSchema() {
+        return schema;
+    }
+
+    public Map<String, String> getOptions() {
+        return options;
+    }
+
+    public List<String> getPartitionKeys() {
+        return partitionKeys;
+    }
+
+    public Optional<String> getComment() {
+        return Optional.ofNullable(comment);
+    }
+
+    // ---------------------------------------------------------------------------------------------
+
+    @Override
+    public String toString() {
+        final String partitionedBy =
+                !partitionKeys.isEmpty()
+                        ? "PARTITIONED BY (" + String.join(", ", partitionKeys) + ")"
+                        : "";
+
+        final String serializedOptions =
+                options.entrySet().stream()
+                        .map(
+                                entry ->
+                                        String.format(
+                                                "  '%s' = '%s'", entry.getKey(), entry.getValue()))
+                        .collect(Collectors.joining(String.format(",%n")));
+
+        return String.format(
+                "[CREATE TABLE … %s%nCOMMENT '%s'%n%s%nWITH (%n%s%n)]",
```

Comment:
nit: I would remove the outer `[` `]` because the `CREATE TABLE` itself is for sure correct.

Reply:
I'm not sure what you mean, but it will not always be valid DDL. We don't escape anything properly: literals, identifiers, … Happy to remove it anyway, though, but I'm afraid it looks too much like it's supposed to return working DDL.